peer_client.go

```diff
@@ -470,40 +464,35 @@ func (c *PeerClient) sendBatch(ctx context.Context, queue []*request) {
 	}
 }
 
-// Shutdown will gracefully shutdown the client connection, until the context is cancelled
-func (c *PeerClient) Shutdown(ctx context.Context) error {
-	// Take the write lock since we're going to modify the closing state
-	c.mutex.Lock()
+// Shutdown waits until outstanding requests have finished
+func (c *PeerClient) Shutdown() {
```

Comment:
Shutdown/Close methods should always accept and `context.Cancel` as there may be some unknown or unforeseen reason the close cannot complete and it then hangs forever. This has happened in production on other services and it's always good to put a limit on how long you will wait for a shutdown to complete. This protects the service from having to be killed by the orchestration system without successfully saving it's internal state.  Shutting down a connection is a low priority thing, if for some reason the workgroup `Wait()`  hangs forever, then we can cancel the shutdown wait, and not interfere with normal shutdown of any service that uses gubernator as a library.

Reply:
I've never seen a shutdown function accept a context. E.g. sql.Close, or connection.Close. 

The context object is meant to be used when there is the potential chance that the request is cancelled by the user. This is not the case, Shutdown is invoked internally, not by a client. We should not use the context as a crutch to fix deadlock issues.

I can refactor this to avoid mutex altogether, but I would like to know if the work would be useful (given the v3 release)

Reply:
And to add, Mailgun plans to abandon this repo in favor of @thrawn01's new repo.  I suppose you can continue using the `/v2` flavor with the new repo until ready to migrate to `/v3`.

Either way, it looks like `/v2` will be deprecated very soon and gRPC support along with it.  I suggest not investing too much time into it.

Reply:
Thanks for the update! I can lend a hand in the new repo 😄

Reply:
> Out of interest, what was the biggest factor for the decision to drop gRPC?

@thrawn01 and I had some big conversations about this.  Reduced complexity and improved performance.  HTTP can perform more efficiently at scale than gRPC if done right.

This would be a good time for @thrawn01 to mention his work on Duh-RPC.

Reply:
> HTTP can perform more efficiently at scale than gRPC if done right.

What's the source for this? My understanding was this this in fact the opposite 😄 https://aws.amazon.com/compare/the-difference-between-grpc-and-rest/

> A gRPC API is better for these use cases:
> High-performance systems
> High data loads
> Real-time or streaming applications

Reply:
Another source: https://github.com/grpc-ecosystem/grpc-gateway#background

> gRPC is great -- it generates API clients and server stubs in many programming languages, it is fast, easy-to-use, **bandwidth-efficient** and its design is combat-proven by Google. However, you might still want to provide a traditional RESTful JSON API as well. Reasons can range from maintaining backward-compatibility, supporting languages or clients that are not well supported by gRPC, to simply maintaining the aesthetics and tooling involved with a RESTful JSON architecture.

Reply:
I intend to publish some more formal information about our GRPC journey at Mailgun. But as a quick overview, we are doing some very high performance and large scale stuff with GRPC and it hasn't lived up to all the hype.

In fact we have discovered that GRPC is slightly slower than using plan old HTTP/1 !!!!!!

As a result, I started working on a project called DUH-RPC, which is still in the works, but you can see the benchmarks comparing HTTP/1 with GRPC here https://github.com/duh-rpc/duh-go-benchmarks

The spec I'm working on is in progress but is here, feed back is welcome. 
https://github.com/duh-rpc/duh-go

I'll be traveling for the next 2 weeks, and I'm not always near my laptop, so I don't expect much movement on these projects until after. I still get notifications for these conversations, so I'll try to reply as I can.